### PR TITLE
Check for `file_types` parameter in the backend

### PR DESCRIPTION
**Pull request by dawoodkhan82**
## Description

Add a file type check to backend for file component.

Closes: #9098

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by dawoodkhan82** [comment]
@yvrjsharma Can you test on windows to see if this fixes the issue? It's expected to show an error popup when an invalid file type is passed in the demo.

**Comment by abidlabs** [comment]
2 quick questions:
* Does this work with the "generic" file_types like "image", "video", "audio", and "text"? 
* Could we add some python unit tests?

**Comment by dawoodkhan82** [comment]
@abidlabs Yes and Yes

File: client/python/gradio_client/utils.py [hunk 1]
```diff
@@ -687,6 +687,24 @@ def get_extension(encoding: str) -> str | None:
...
+def is_valid_file(mime_type: str, file_types: list[str]) -> bool:
+    for file_type in file_types:
+        if file_type == "file":
+            return True
+        if file_type.startswith("."):
+            file_type = file_type.lstrip(".").lower()
+            mime_type_split = mime_type.lower().split("/")
+            if (
+                file_type in mime_type_split
+            ):  # Check if the MIME type matches the file extension
```
Comment (abidlabs) [inline client/python/gradio_client/utils.py:699]: Should this be:
```suggestion
            mime_type_split = mime_type.lower().split("/")
            if file_type == mime_type_split[1]
```

File: client/python/gradio_client/utils.py [hunk 2]
```diff
@@ -687,6 +687,24 @@ def get_extension(encoding: str) -> str | None:
...
+                return True
+        elif mime_type.startswith(
+            f"{file_type}/"
+        ):  # Check if the MIME type starts with the file type
```
Comment (abidlabs) [inline client/python/gradio_client/utils.py:703]:
```suggestion
        ): 
```

File: gradio/components/file.py [hunk 3]
```diff
@@ -125,6 +126,15 @@ def __init__(
...
         file_name = f.path
         if self.type == "filepath":
+            mime_type = client_utils.get_mimetype(file_name)
+            if (
+                self.file_types
+                and mime_type
+                and not client_utils.is_valid_file(mime_type, self.file_types)
+            ):
+                raise Error(
+                    f"Invalid file type: {mime_type}. Please choose from: {self.file_types}"
```
Comment (abidlabs) [inline gradio/components/file.py:136]:
```suggestion
                    f"Invalid file type: {mime_type}. Please upload a file that is one of these formats: {self.file_types}"
```

**Comment by abidlabs** [comment]
Works great @dawoodkhan82! I would just add this same logic to the other components that support `file_types`: `Gallery`, `MultimodalTextbox`, `UploadButton`

File: gradio/components/file.py [hunk 4]
```diff
@@ -125,6 +126,15 @@ def __init__(
...
         if self.type == "filepath":
+            mime_type = client_utils.get_mimetype(file_name)
+            if (
+                self.file_types
+                and mime_type
+                and not client_utils.is_valid_file(mime_type, self.file_types)
+            ):
+                raise Error(
+                    f"Invalid file type: {mime_type}. Please upload a file that is one of these formats: {self.file_types}"
+                )
```
Comment (abidlabs) [inline gradio/components/file.py:137]: Looks good @dawoodkhan82. nit, but there is some code duplication, thoughts on abstracting this part into a function and calling that function from the four components?

Comment (abidlabs) [reply gradio/components/file.py:137]: Or I guess you could even just move some of this into `is_valid_file`?

Comment (abidlabs) [reply gradio/components/file.py:137]: Concretely, I'd suggest moving the `mime_type = client_utils.get_mimetype(file_name)` into the `is_valid_file` function to reduce the duplication

Comment (dawoodkhan82) [reply gradio/components/file.py:137]: Yeah good point

File: client/python/test/test_utils.py [hunk 5]
```diff
@@ -71,6 +71,21 @@ def test_decode_base64_to_file():
     assert isinstance(temp_file, tempfile._TemporaryFileWrapper)
 
 
+@pytest.mark.parametrize(
```
Comment (abidlabs) [inline client/python/test/test_utils.py:74]: Code looks good, we'll need to just update these tests

**Review by abidlabs — APPROVED** [review]
code lgtm, should be good once tests are fixed!

**Review by abidlabs — APPROVED** [review]
All the other failing python tests are unrelated, so this looks good to merge!

**Pull request merged**